### PR TITLE
Add API for read barrier

### LINKED ISSUE #19213: RFE: Add API for read barrier
This could be useful for Scylla Manager in the backup procedure. It contains the 3 following steps:
- snapshot all requested tables
- await schema agreement
- backup the output of DESCRIBE SCHEMA WITH INTERNALS

As SM wants to backup only a single version of DESC SCHEMA (backing up DESC SCHEMA from every node creates decision problem with restore procedure), it queries DESC SCHEMA on only 1 of the nodes. The problem is that SM needs to backup schema that is not older than sstables in the snapshot - older sstables should work fine with altered schema, but newer sstables used with older schema might create problems. The idea was to await schema agreement between snapshot and DESC SCHEMA stages, so that all nodes have the same schema - meaning that all nodes have schema that's not older than snapshot sstables.

Problem with awaiting schema agreement is that it is usually done from a single node POV. Gocql implementation does that by checking if schema versions from `system.local/peers` are the same, but it does so only from the control connection POV. From the backup perspective it is not enough. See the following scenario:
- node1 gets schema change
- node1 gets inserts with new schema
- SM backup is performed, assume that gocql control connection is connected to node2
- from node2 POV, cluster is in a state of schema agreement (the schema change wasn't propagated to it yet)
- SM backs up DESC SCHEMA output from node2

This scenario (if possible) results in having newer sstables than schema in the backup and that might create problem.
On the other hand, if we replaced await schema agreement with read barrier followed by querying DESC SCHEMA on the same node, on which read barrier was performed, then it should work fine.

Right now read barrier API is not exposed, but @kbr-scylla mentioned that it can be simulated by running some dummy schema change:
```
async def read_barrier(cql: Session, host: Host):
    """To issue a read barrier it is sufficient to attempt dropping a
    non-existing table. We need to use `if exists`, otherwise the statement
    would fail on prepare/validate step which happens before a read barrier is
    performed.
    """
    await cql.run_async("drop table if exists nosuchkeyspace.nosuchtable", host = host)
```

Questions:
1. Does read carrier indeed solve SM issue? 
2. Is this enough of a reason to introduce read barrier in Scylla API?
3. Is using the workaround of sending dummy schema change safe and will it be supported in the future?

Of course, there are other ways to make sure that backed up DESC SCHEMA is newer than snapshot sstables. SM could probably do it by:
- implementing custom host selection policy for the driver - not sure if it would be easy outside of gocql pkg - and using it to query `system.local` on every node
- creating many single host sessions and querying `system.local` from every host manually
- changing gocql to somehow support that in `AwaitSchemaAgreement` method
- in case `schema_version` contains timestamp information, SM could parse it in order to find a node with the newest schema and use it to query DESC SCHEMA

What do you think about that?
@piodul @kbr-scylla @kostja @karol-kokoszka

### PR DESCRIPTION
**Pull request by emdroid**
Introduce REST API for triggering a read barrier.

This is to make sure the database schema is up to date on the node where the read barrier is triggered. One of the use cases is the database backup via the Scylla Manager, which requires that the schema backed up is matching the data or newer (data can be migrated, but an older schema would cause issues).

Fixes #19213

### CODE REVIEW

File: test/topology_custom/test_table_desc_read_barrier.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    logging.info("Creating a table")
+    random_tables = RandomTables(request.node.name, manager, unique_name(), 1)
+    table = await random_tables.add_table(ncolumns=5, pks=1)
+
+    s1, s2, *_ = servers
+
+    async with inject_error(manager.api, s1.ip_addr, 'group0_state_machine::apply',
+                            parameters={"second_node_ip": s2.ip_addr}) as handler:
+        logger.info("Altering table")
+        cql.max_schema_agreement_wait = None
```
[the remark this answers is not in the document]
Comment (kbr-scylla) [reply test/topology_custom/test_table_desc_read_barrier.py:56]: From driver docs
https://docs.datastax.com/en/developer/python-driver/3.18/api/cassandra/cluster/index.html
> The maximum duration (in seconds) that the driver will wait for schema agreement across the cluster. Defaults to ten seconds. If set <= 0, the driver will bypass schema agreement waits altogether.

Does `None` give the same effect as `0`?

[the remark this answers is not in the document]
Comment (kbr-scylla) [reply test/topology_custom/test_table_desc_read_barrier.py:56]: Checked where it's used in the driver
https://github.com/scylladb/python-driver/blob/11d3499351373c27bc39cff85b4cea393b67b7a8/cassandra/cluster.py#L4125


```python3
    def wait_for_schema_agreement(self, connection=None, preloaded_results=None, wait_time=None):

        total_timeout = wait_time if wait_time is not None else self._cluster.max_schema_agreement_wait
        if total_timeout <= 0:
            return True

...

            while elapsed < total_timeout:
                peers_query = QueryMessage(query=select_peers_query, consistency_level=cl)
                local_query = QueryMessage(query=self._SELECT_SCHEMA_LOCAL, consistency_level=cl)
                try:
                    timeout = min(self._timeout, total_timeout - elapsed)

```

It looks like `None` will not give sensible results

[the remark this answers is not in the document]
Comment (emdroid) [reply test/topology_custom/test_table_desc_read_barrier.py:56]: I have seen `None` somewhere, is probably the same (seeing no difference in the test run), but will be better to use `0` to be in line with the docs.

File: test/topology_custom/test_table_desc_read_barrier.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        - create a table
+        - inject error (sleep) on the first node
+        - disable the schema agreement wait
+    Act:
+        - alter the table (change schema)
+        - trigger the read barrier
+    Assert:
+        - the schema retrieved on both hosts matches
+    """
+    servers = [await manager.server_add() for _ in range(2)]
```
Comment (kbr-scylla) [inline test/topology_custom/test_table_desc_read_barrier.py:41]: You can boot them concurrently (using `servers_add`), might be a bit faster

File: test/topology_custom/test_table_desc_read_barrier.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+    logging.info("Waiting until driver connects to every server")
+    cql = manager.get_cql()
+    h1, h2, *_ = await wait_for_cql_and_get_hosts(cql, servers, time.time() + 60)
+
+    logging.info("Creating a table")
+    random_tables = RandomTables(request.node.name, manager, unique_name(), 1)
+    table = await random_tables.add_table(ncolumns=5, pks=1)
```
Comment (kbr-scylla) [inline test/topology_custom/test_table_desc_read_barrier.py:49]: I'd recommend just writing down the CQL statement with some predefined simple schema instead of using the `RandomTables` utility which in this case only obfuscates the test a bit IMO.

File: service/raft/group0_state_machine.cc [hunk 4]
```diff
@@ -184,6 +184,12 @@ future<> group0_state_machine::merge_and_apply(group0_state_machine_merger& merg
 future<> group0_state_machine::apply(std::vector<raft::command_cref> command) {
     slogger.trace("apply() is called with {} commands", command.size());
 
+    co_await utils::get_local_injector().inject("group0_state_machine::apply", [](auto& handler) -> future<> {
```
Comment (gleb-cloudius) [inline service/raft/group0_state_machine.cc:187]: Please use better name for injection point that also describes what it does (sleeps). Also a nit: there should be a space between [] and (.

Comment (emdroid) [reply service/raft/group0_state_machine.cc]: Will "group0_state_machine::delay_apply" be a good name? Also fixed the lambda, in fact I found another example of the timeout and it looks like I don't need to use lambda at all, just to provide the delay as duration.

(the space between [] and ( has been removed by the clang-format that I tried to setup ... looked into the options, but apparently there is no option in the clang-format that would keep the space in between these :disappointed: )

Comment (gleb-cloudius) [reply service/raft/group0_state_machine.cc]: > Will "group0_state_machine::delay_apply" be a good name?

Yes.

> Also fixed the lambda, in fact I found another example of the timeout and it looks like I don't need to use lambda at all, just to provide the delay as duration.

Right.

**Comment by emdroid** [comment]
New version:
- added the optional timeout parameter to the "read_barrier" API request
- handling it as a value for the raft_timeout (if not set, using the default value)

**Comment by nyh** [comment]
A general note:

Before creating a pull request, it's good practice to create an issue (which the pull-request "Fixes: ..."), to explain what the problem is before you explain what the solution here. In particular, what is a "read barrier" when when does one need to wait for it? Is the **REST API** the appropriate place to put this read barrier command, or maybe CQL users need it and it should be a CQL command? Or maybe both? An issue will be a place to explain and/or discuss the needs of this feature - separately from the specific implementation code.

**Comment by nyh** [comment]
> A general note:
> 
> Before creating a pull request, it's good practice to create an issue (which the pull-request "Fixes: ..."), to explain what the problem is before you explain what the solution here. In particular, what is a "read barrier" when when does one need to wait for it? Is the **REST API** the appropriate place to put this read barrier command, or maybe CQL users need it and it should be a CQL command? Or maybe both? An issue will be a place to explain and/or discuss the needs of this feature - separately from the specific implementation code.

Oh, now I see there is already an issue: https://github.com/scylladb/scylladb/issues/19213
So you just forgot to put "Fixes #19213" on this pull request.

File: test/topology_custom/test_table_desc_read_barrier.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    await cql.run_async("create table ks.t (pk int primary key)")
+
+    logger.info("Disabling the schema agreement wait")
+    assert hasattr(cql.cluster, "max_schema_agreement_wait")
+    cql.cluster.max_schema_agreement_wait = 0
+
+    async with inject_error(manager.api, servers[0].ip_addr, 'group0_state_machine::delay_apply'):
+        logger.info("Altering table")
+        await cql.run_async("alter table ks.t add s1 int", host=hosts[1])
+        # TODO: trigger the read barrier once available
```
Comment (nyh) [inline test/topology_custom/test_table_desc_read_barrier.py]: Does this test pass at this stage, with this TODO?
If it doesn't, it breaks bisectability.
Maybe you should just add the test, in its complete form, as the second patch and not the first?

File: test/topology_custom/test_table_desc_read_barrier.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+async def test_table_desc_read_barrier(request, manager: ManagerClient) -> None:
+    """
+    Regression test for #19213.
+
+    Test that the schema is always updated after read barrier.
+
+    Arrange:
+        - start 2 nodes
+        - create a table
+        - inject error (sleep) on the first node
```
Comment (nyh) [inline test/topology_custom/test_table_desc_read_barrier.py:31]: Maybe we call this function inject_error(), but you are injecting a sleep, not an error.
It can be useful to say in a couple of words *where* in the server code you inject this sleep (before completing something?).

File: test/topology_custom/test_table_desc_read_barrier.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        - trigger the read barrier
+    Assert:
+        - the schema retrieved on both hosts matches
+    """
+    servers = await manager.servers_add(2)
+
+    logging.info("Waiting until driver connects to every server")
+    cql = manager.get_cql()
+    # this is needed so that the hosts will be matching the servers
+    hosts = [(await wait_for_cql_and_get_hosts(cql, [s], time.time() + 60))[0] for s in servers]
```
Comment (nyh) [inline test/topology_custom/test_table_desc_read_barrier.py:44]: @kbr-scylla is this "wait_for_cql_and_get_hosts" necessary? Can't the test code assume that after it calls servers_add(2) and get_cql(), it actually has access to both nodes in CQL requests?

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 5 above)
Comment (emdroid) [reply test/topology_custom/test_table_desc_read_barrier.py]: The test passes (so it should not break bisectability) - the expectation is that the schema is different without the read barrier (which the test checks here), after adding the barrier the assertion is switched for equality

(initially it was just switching the condition to equality after adding the barrier, however later I actually kept the first check for unequality to make sure the initial assumption holds, and then triggering the read barrier and check for equality of the schemas)

But it might indeed make sense to switch the order (or even squash both to one commit).

File: api/raft.cc [hunk 8]
```diff
@@ -74,12 +74,32 @@ void set_raft(http_context&, httpd::routes& r, sharded<service::raft_group_regis
...
+            raft::group_id gid{utils::UUID{req->get_query_param("group_id")}};
+            return raft_gr.local().get_server_with_timeouts(gid);
+        });
+        auto timeout = std::invoke([timeout_str = req->get_query_param("timeout")] {
+            if (timeout_str.empty()) {
+                return ::service::raft_timeout{};
+            }
+            auto dur = std::stoll(timeout_str);
+            if (dur <= 0) {
+                throw std::runtime_error{"Timeout must be a positive number."};
```
Comment (nyh) [inline api/raft.cc:91]: I see that other code in api/ throws bad_param_exception when they encounter bad parameters, not the generic std::runtime_error. Please consider if this code should do it too.

File: test/topology_custom/test_table_desc_read_barrier.py [hunk 9]
```diff
@@ -55,7 +55,15 @@ async def test_table_desc_read_barrier(request, manager: ManagerClient) -> None:
     async with inject_error(manager.api, servers[0].ip_addr, 'group0_state_machine::delay_apply'):
         logger.info("Altering table")
         await cql.run_async("alter table ks.t add s1 int", host=hosts[1])
-        # TODO: trigger the read barrier once available
-        desc_schema = [await cql.run_async("DESC SCHEMA", host=h) for h in hosts]
 
+        # verify that there is a schema difference
+        # (raft apply delayed, so the first server doesn't have the new schema yet)
+        desc_schema = [await cql.run_async("DESC SCHEMA", host=h) for h in hosts]
```
Comment (nyh) [inline test/topology_custom/test_table_desc_read_barrier.py:61]: This test has the potential of becoming flaky if the sleep (I think I remember it is 500 milliseconds) ends before this code gets to run - perhaps because of some super-slow test machine.

I'm not sure what to do about this. Ideally instead of a sleep injection you'll have injection that "hangs" forever and a second injection which releases this hang. But maybe it's not easy to do.

File: test/topology_custom/test_table_desc_read_barrier.py [hunk 10]
```diff
@@ -55,7 +55,15 @@ async def test_table_desc_read_barrier(request, manager: ManagerClient) -> None:
...
         assert desc_schema[0] != desc_schema[1]
+
+        # waits for the schema fully updated on all nodes (after the delay ends)
```
Comment (nyh) [inline test/topology_custom/test_table_desc_read_barrier.py:64]: @kbr-scylla @gleb-cloudius is this comment true - does a read barrier wait for **all** nodes? Isn't it just a quorum?

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 7 above)
Comment (emdroid) [reply test/topology_custom/test_table_desc_read_barrier.py:44]: I still need the hosts at some point in the test (and in particular, in the same order as servers, or at least one of them to match) ... although previously I needed both, now I mostly just need the whole list, so perhaps I could use `cql.cluster.metadata.all_hosts()` (not sure about the ordering though)

File: test/topology_custom/test_table_desc_read_barrier.py [hunk 11]
```diff
@@ -55,7 +55,15 @@ async def test_table_desc_read_barrier(request, manager: ManagerClient) -> None:
...
+        await read_barrier(manager.api, servers[0].ip_addr)
+
+        # verify that there is no schema difference after the read barrier
+        desc_schema = [await cql.run_async("DESC SCHEMA", host=h) for h in hosts]
```
Comment (nyh) [inline test/topology_custom/test_table_desc_read_barrier.py:68]: I have to admit I got confused here. Since "DESC SCHEMA" is an official CQL operation that queries, well,..., the schema, isn't its responsibility to wait until the node is sure about the schema? Why doesn't DESC SCHEMA do its own "read barrier" if that is needed to make read of the schema?

File: api/raft.cc (hunk 8 above)
Comment (emdroid) [reply api/raft.cc:91]: Yeah, this was copied from the other request slightly above, but `bad_param_exception` might make more sense.

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 10 above)
Comment (gleb-cloudius) [reply test/topology_custom/test_table_desc_read_barrier.py:64]: A read barrier only waits for the node it is executed on to see the latest state.

Comment (emdroid) [reply test/topology_custom/test_table_desc_read_barrier.py:64]: In fact it should only wait for the schema to be updated on the node where the read barrier is issued (i.e. on the first node) - I assume at least

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 11 above)
Comment (emdroid) [reply test/topology_custom/test_table_desc_read_barrier.py:68]: As I understand according to the explanation in #19213 it only does "schema agreement wait", which is only done locally and is what causes the problem with old schema version, as explained here:

> Problem with awaiting schema agreement is that it is usually done from a single node POV. Gocql implementation does that by checking if schema versions from system.local/peers are the same, but it does so only from the control connection POV

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 10 above)
Comment (kbr-scylla) [reply test/topology_custom/test_table_desc_read_barrier.py:64]: After read barrier successfully finishes on node X, it is guaranteed that node X has applied all group 0 commands that were committed before this read barrier started. (And maybe more if concurrently more commands were being committed.)

Read barrier implementation does require contacting a quorum (the impl RPCs the leader and then the leader RPCs a quorum -- roughly speaking) but it doesn't say anything about whether this quorum will apply the commands, only node X.

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 9 above)
Comment (emdroid) [reply test/topology_custom/test_table_desc_read_barrier.py:61]: Yes, I was thinking about it as well, the tricky part is that the wait would need to be unblocked after the read barrier is triggered and waiting (so basically a deadlocking scenario) ... when testing locally with 100 reps and pool size 10 and 20 the test was always 100% passing (in case of flackiness we'd probably need to increase the timeout).

Will still give it a thought, if I can see some way to make it work fully deterministically.

File: api/api-doc/raft.json [hunk 12]
```diff
@@ -62,6 +62,38 @@
                ]
             }
          ]
+      },
+      {
+         "path": "/raft/read_barrier",
+         "operations": [
+            {
+               "method": "POST",
+               "summary": "Triggers read barrier for the given Raft group",
```
Comment (avikivity) [inline api/api-doc/raft.json:71]: Maybe add some words on when this is useful?

**Comment by emdroid** [comment]
New version:
- improved the rest api summary
- using `bad_param_exception` instead of `std::runtime_error`
- improvements and cleanup of the test according to the review comments
- remove the potential of flakiness from the test (removed the initial check for different schemas before the read barrier)

File: api/api-doc/raft.json (hunk 12 above)
Comment (emdroid) [reply api/api-doc/raft.json]: Updated the summary

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 10 above)
Comment (emdroid) [reply test/topology_custom/test_table_desc_read_barrier.py]: fixed the comment

File: api/raft.cc (hunk 8 above)
Comment (emdroid) [reply api/raft.cc]: changed to `bad_param_exception`

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 6 above)
Comment (emdroid) [reply test/topology_custom/test_table_desc_read_barrier.py]: improved the comment

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 5 above)
Comment (emdroid) [reply test/topology_custom/test_table_desc_read_barrier.py]: changed to comparison for equality (as is expected later) and marked the test `xfail` in this commit (that makes it "pass as failed"), the `xfail` is then removed in the impl commit

in particular, the test has been made first to make sure the issue is reproduced first (to ensure that the read barrier addition actually fixes the problem, kind of in a TDD sense)

**Comment by emdroid** [comment]
New version:
- make the test more deterministic by using a handler message to resume the raft apply

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 9 above)
Comment (emdroid) [reply test/topology_custom/test_table_desc_read_barrier.py:61]: I managed to use handler message to resume after calling the read barrier and still avoid the deadlock, by using the `asyncio.gather()` to unpause in parallel. There is still little bit if non-determinism in the sense that the resume might happen before the read barrier (added an extra small sleep before resume to make that less likely), but the test will still succeed in both cases (thus it should not be flaky in any case).

File: api/raft.cc [hunk 13]
```diff
@@ -74,12 +74,33 @@ void set_raft(http_context&, httpd::routes& r, sharded<service::raft_group_regis
...
+            if (timeout_str.empty()) {
+                return ::service::raft_timeout{};
+            }
+            auto dur = std::stoll(timeout_str);
+            if (dur <= 0) {
+                throw bad_param_exception{"Timeout must be a positive number."};
+            }
+            return ::service::raft_timeout{.value = lowres_clock::now() + std::chrono::seconds{dur}};
+        });
+        co_await srv.read_barrier(nullptr, timeout);
```
Comment (tgrabiec) [inline api/raft.cc:95]: @kbr-scylla Doesn't group0 server live only on shard 0?

Comment (kbr-scylla) [reply api/raft.cc:95]: Yes

Comment (kbr-scylla) [reply api/raft.cc:95]: So this whole thing needs to happen on the correct shard for the requested group (and if no group requested -- on shard 0 for group 0)

For example similar API, `trigger_snapshot`, uses this:
```cpp
        co_await raft_gr.invoke_on_all([gid, timeout_dur, &found_srv] (service::raft_group_registry& raft_gr) -> future<> {
            auto* srv = raft_gr.find_server(gid);
            if (!srv) {
                co_return;
            }

```

by calling `invoke_on_all` for the `raft_group_registry` service, all shards but one (the one containing the group) will do nothing

File: test/topology_custom/test_table_desc_read_barrier.py [hunk 14]
```diff
@@ -60,9 +61,17 @@ async def test_table_desc_read_barrier(manager: ManagerClient) -> None:
...
+            # extra sleep before resuming the apply
+            # (keep paused for some more time while read barrier is triggered)
+            await asyncio.sleep(timedelta(milliseconds=100).total_seconds())
+            await handler.message()
+
+        # wait for the first node to see the latest state (after the delay ends)
+        await asyncio.gather(
+            read_barrier(manager.api, servers[0].ip_addr),
+            resume()
+        )
```
[the remark this answers is not in the document]
Comment (kbr-scylla) [reply test/topology_custom/test_table_desc_read_barrier.py:74]: I don't see the necessity for using the handler version of `inject` at all.

Just do a sleep injection. In the test call read_barrier; it will wait for sleep to finish because it will wait for `apply` to finish.

[the remark this answers is not in the document]
Comment (kbr-scylla) [reply test/topology_custom/test_table_desc_read_barrier.py:74]: > I think this test is fine and we don't need a perfect deterministic solution. However, we have ways to achieve it.

In this case I wouldn't waste time trying to come up with a fully deterministic solution

I would remove the assertion
```
         assert desc_schema[0] != desc_schema[1]
```
since this is not what we want to test at the end. We want to test that the schema is synced after the read barrier, we don't want to test that the schema is not synced before it.

And then a simple sleep will work without making the test flaky. It might make it less accurate -- in theory the change may propagate naturally and test would pass without the barrier, but in most cases the barrier is still required, and that is enough to test that the barrier works.

File: api/api-doc/raft.json [hunk 15]
```diff
@@ -62,6 +62,38 @@
                ]
             }
          ]
+      },
+      {
+         "path": "/raft/read_barrier",
+         "operations": [
+            {
+               "method": "POST",
+               "summary": "Triggers read barrier for the given Raft group to ensure the schema is on the latest agreed state",
```
Comment (kbr-scylla) [inline api/api-doc/raft.json:71]: schema is specific for group 0. And even for group 0, there are other things living in it than just schema.
So the API is more generic than waiting for latest schema.

Perhaps you can still mention schema but make it clear that it's just an example. I.e. "for the given Raft group. For example, you can use it with group 0 to ensure that the schema ..."

Also, "latest" is ambiguous, because new commands can be committed concurrently while the barrier is in progress, and it might or might not see them. What does "latest" mean in that case?

File: api/raft.cc [hunk 16]
```diff
@@ -56,7 +56,7 @@ void set_raft(http_context&, httpd::routes& r, sharded<service::raft_group_regis
         });
 
         if (!found_srv) {
-            throw std::runtime_error{fmt::format("Server for group ID {} not found", gid)};
+            throw bad_param_exception{fmt::format("Server for group ID {} not found", gid)};
```
Comment (kbr-scylla) [inline api/raft.cc:59]: This is an unrelated change. It's nice, because you want to make the APIs consistent -- but please put it in a separate commit to make it clear that it's not part of implementing read barrier.

Comment (kbr-scylla) [reply api/raft.cc:59]: (The commit should probably come before this one, so in this one you can use `bad_param_exception` without introducing inconsistencies)

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 11 above)
Comment (kbr-scylla) [reply test/topology_custom/test_table_desc_read_barrier.py:77]: IIUC `DESC SCHEMA` does not do any wait. The aforementioned wait is something that driver does after performing a schema **change**, but as the issue explains, it is not enough for the backup use case.

Why doesn't `DESC SCHEMA` do its own barrier? Well, it never did, it was always returning local schema. And I think that's fine; we might sometimes want to check local schema of a node without forcing it to communicate with quorum. That would be a backwards-incompatible interface change and potentially eliminate some valid use cases. If you want to ensure you obtain recent schema, you can call read_barrier explicitly. I think that's fine.

**Comment by kbr-scylla** [comment]
Please remove `RFE: ` prefix from PR title, the issue you're fixing is indeed a Request For Enhancement, but the pull request is not; it is the answer to that request.

File: api/api-doc/raft.json [hunk 17]
```diff
@@ -62,6 +62,38 @@
                ]
             }
          ]
```
[the remark this answers is not in the document]
Comment (kbr-scylla) [reply api/api-doc/raft.json:64]: One potential problem with clean-up is that it might cause CI to run 100 repetitions of multiple tests, killing the run (or timing it out).

So we might need to do the follow-up in batches -- and multiple PRs will be required in that case, or perhaps somehow bypassing CI / making it run fewer repetitions.

To make the work more incremental, I'd actually prefer separate PR(s) for this follow-up.

**Comment by emdroid** [comment]
New version:
* rebased
* simplified the test and error injection back to the original solution (the test will not become flaky even if the injected raft apply pause ends earlier than the read barrier occasionally, because the test will succeed in both cases)
* fixed to make the code run on the correct shard(s)
* updated the API description
* migrated the `trigger_snapshot` to also use the `raft_timeout` (separate commit)

File: api/raft.cc (hunk 13 above)
Comment (emdroid) [reply api/raft.cc]: fixed the code to run on the correct shards

File: test/topology_custom/test_table_desc_read_barrier.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (emdroid) [reply test/topology_custom/test_table_desc_read_barrier.py]: reverted to the previous version to just use the simple pause (increase the timeout to 1s)

File: api/raft.cc (hunk 16 above)
Comment (emdroid) [reply api/raft.cc:68]: moved to a separate commit

File: api/raft.cc [hunk 18]
```diff
@@ -83,12 +83,40 @@ void set_raft(http_context&, httpd::routes& r, sharded<service::raft_group_regis
             return json_return_type(srv.current_leader().to_sstring());
         });
     });
+    r::read_barrier.set(r, [&raft_gr] (std::unique_ptr<http::request> req) -> future<json_return_type> {
+        auto timeout = get_request_timeout(*req);
+
+        if (!req->query_parameters.contains("group_id")) {
+            // Read barrier on group 0 by default
+            co_await raft_gr.invoke_on(0, [timeout] (service::raft_group_registry& raft_gr) -> future<> {
+                co_await raft_gr.group0_with_timeouts().read_barrier(nullptr, timeout);
```
Comment (kbr-scylla) [inline api/raft.cc:92]: nit: pessimizing coroutine (by using `return` here instead of `co_await`, we'd avoid one allocation, without losing any readability or anything).

Doesn't matter though, this is not a hot path, so we can keep it like that. (But it would potentially be important for hot path code -- good to remember in the future.)

**Review by kbr-scylla — APPROVED** [review]
Very nice. Thanks for the cleanups too.

**Comment by emdroid** [comment]
New version:
* minor cleanup (const correctness, returns)

**Comment by kbr-scylla** [comment]
Qd

File: api/api-doc/raft.json (hunk 17 above)
[the remark this answers is not in the document]
Comment (emdroid) [reply api/api-doc/raft.json:64]: Created the #19662 for the follow-up cleanup.
